### PR TITLE
[cxx-interop] Runtime support for foreign reference types.

### PR DESCRIPTION
**Pull request by zoecarver**
This is the first of a few PRs to add a value witness table, metadata, and runtime support for foreign reference types.

### CODE REVIEW

File: include/swift/ABI/MetadataKind.def [hunk 1]
```diff
@@ -53,6 +53,8 @@ NOMINALTYPEMETADATAKIND(Optional, 2 | MetadataKindIsNonHeap)
 /// A foreign class, such as a Core Foundation class.
 METADATAKIND(ForeignClass, 3 | MetadataKindIsNonHeap)
 
+METADATAKIND(ForeignReferenceType, 4 | MetadataKindIsNonHeap)
```
Comment (zoecarver) [inline include/swift/ABI/MetadataKind.def:56]: @rjmccall I have no idea what this value should be.

File: lib/IRGen/GenMeta.cpp [hunk 2]
```diff
@@ -1688,7 +1688,7 @@ namespace {
         VTable(IGM.getSILModule().lookUpVTable(getType())),
         Resilient(IGM.hasResilientMetadata(Type, ResilienceExpansion::Minimal)) {
 
-      if (getType()->isForeign() || Type->isForeignReferenceType())
```
Comment (zoecarver) [inline lib/IRGen/GenMeta.cpp:1691]: (This was just a redundant check.)

File: lib/IRGen/GenMeta.cpp [hunk 3]
```diff
@@ -1788,7 +1787,8 @@ namespace {
       // is a foreign class.
       if ((IGM.IRGen.Opts.ReflectionMetadata !=
            ReflectionMetadataMode::Runtime) ||
-          getType()->isForeign()) {
+          getType()->isForeign() ||
+          getType()->isForeignReferenceType()) {
```
Comment (zoecarver) [inline lib/IRGen/GenMeta.cpp:1791]: I should remove this (see above).

**Comment by zoecarver** [comment]
@rjmccall can you verify that this won't affect ABI at all for existing metadata? (Specifically with the added METADATAKIND.)

**Comment by zoecarver** [comment]
@swift-ci please test.

File: lib/IRGen/GenMeta.cpp [hunk 4]
```diff
@@ -5532,6 +5518,60 @@ namespace {
...
+  /// A builder for ForeignReferenceTypeMetadata.
+  class ForeignReferenceTypeMetadataBuilder :
+      public ForeignMetadataBuilderBase<ForeignReferenceTypeMetadataBuilder,
+                                        ForeignClassMetadataBuilderBase<ForeignReferenceTypeMetadataBuilder>> {
+  public:
+    ForeignReferenceTypeMetadataBuilder(IRGenModule &IGM, ClassDecl *target,
+                                        ConstantStructBuilder &B)
+        : ForeignMetadataBuilderBase(IGM, target, B) {
+      assert(getTargetType()->isForeignReferenceType() &&
+             "foreign reference type metadata build must be used on foreign reference types.");
```
Comment (rjmccall) [inline lib/IRGen/GenMeta.cpp:5531]: Please add the inverse of this assertion in `ForeignClassMetadataBuilder`.

File: lib/IRGen/GenMeta.cpp [hunk 5]
```diff
@@ -5532,6 +5518,60 @@ namespace {
...
+      B.addSignedPointer(descriptor,
+                         IGM.getOptions().PointerAuth.TypeDescriptors,
+                         PointerAuthEntity::Special::TypeDescriptor);
+    }
+
+    void addSuperclass() {
+      // Always leave the superclass pointer unfilled.  We'll have to
+      // unique it during initialization anyway, so we might as well spare
+      // ourselves the load-time work.
+      B.addNullPointer(IGM.TypeMetadataPtrTy);
```
Comment (rjmccall) [inline lib/IRGen/GenMeta.cpp:5565]: I don't see this superclass field in the metadata class.

In general, if there isn't any commonality of layout in the metadata classes, I'm not sure we should be using related metadata builder classes; it just ends up obscuring how the code in GenMeta matches the structs in Metadata.h.

Comment (zoecarver) [reply lib/IRGen/GenMeta.cpp:5565]: This is using `ForeignMetadataBuilderBase` which seemed generic enough. I'll just make this into a no-op + assert. Let me know if you want me to add another builder, though.

**Comment by zoecarver** [comment]
@swift-ci please test.

File: test/Interop/Cxx/foreign-reference/witness-table.swift [hunk 6]
```diff
@@ -1,7 +1,7 @@
 // RUN: %target-run-simple-swift(-I %S/Inputs/ -Xfrontend -enable-experimental-cxx-interop -Xfrontend -validate-tbd-against-ir=none -Xfrontend -disable-llvm-verify -g)
 //
 // REQUIRES: executable_test
-// REQUIRES: foreign-reference-type-metadata-support
+// XFAIL: OS=windows-msvc
```
Comment (zoecarver) [inline test/Interop/Cxx/foreign-reference/witness-table.swift:4]: This matches all existing executable tests for foreign reference types.

File: lib/IRGen/GenMeta.cpp (hunk 5 above)
Comment (zoecarver) [reply lib/IRGen/GenMeta.cpp]: Looking at this again, I see that I am actually using `ForeignClassMetadataBuilderBase`. I added my own builder for FRTs and updated it to use that. Thanks for pointing this out.

**Comment by zoecarver** [comment]
@swift-ci please test.

**Comment by zoecarver** [comment]
@swift-ci please test macOS

**Comment by zoecarver** [comment]
I'm going to go ahead and merge this. If there are any more comments, I'm happy to address them post-commit (or if necessary, revert).

File: stdlib/public/runtime/FoundationSupport.cpp [hunk 7]
```diff
@@ -31,6 +31,7 @@ swift::_swift_isObjCTypeNameSerializable(Class theClass) {
   switch (type->getKind()) {
   case MetadataKind::ObjCClassWrapper:
   case MetadataKind::ForeignClass:
+  case MetadataKind::ForeignReferenceType:
```
Comment (etcwilde) [inline stdlib/public/runtime/FoundationSupport.cpp:34]: Is this right?
- Zoe wrote this from the future

**Comment by etcwilde** [comment]
The pieces in the runtime directory will need to be back deployed.

**Pull request merged**